### ISSUE TITLE
Figure out a testing story.

### ISSUE THREAD

**Issue by emilio** [fiji-flo/bikeshed-rs#4]
We should figure out how do we want bikeshed-rs to be tested.

This probably involves:

 * Looking into how upstream bikeshed is tested. If we can come up with a way to share the tests that'd be best.

 * Decide whether we want to use regular `#[test]` functions, or something more like what `cbindgen` and `bindgen` use, where `build.rs` basically reads a `tests/` folder and auto-generates the relevant cargo tests: https://github.com/eqrion/cbindgen/blob/master/build.rs

**Comment by annevk** [comment]
See https://github.com/tabatkins/bikeshed/tree/master/tests. The idea is, given an input .bs, does it generate the expected .html.

**Comment by whichxjy** [comment]
I use the [serialize](https://docs.rs/kuchiki/0.8.0/kuchiki/struct.NodeRef.html#method.serialize) method of `kuchiki` to serialize the DOM tree now, but there's [a special HTML serializer](https://github.com/tabatkins/bikeshed/blob/56380d92311d15130247d5db916cacaa3b1297f8/bikeshed/HTMLSerializer.py#L9) in `bikeshed`. If we want to use the expected HTML files in `bikeshed` for testing, maybe we need to write a serializer like that.

**Comment by emilio** [comment]
Alternatively, can we parse the bikeshed HTML with kuchiki, and serialize it back to guarantee it's idempotent?

Output doesn't have  to be whitespace-compatible IMO.

**Comment by whichxjy** [comment]
That makes sense if there's nothing special about the serializer in `bikeshed`. What about building DOM trees from source files and expected HTML files, and then comparing them without serialization?

**Comment by annevk** [comment]
That's also fine, but you'll still run into the whitespace issues I suspect. And potentially attribute ordering issues as well. (Attribute order is unfortunately still implementation-defined.)

**Comment by emilio** [comment]
Yeah, however those issues should be easy to work around / ignore in the comparison (sorting the attributes, and skipping child whitespace text nodes where appropriate).

So it seems pretty reasonable to me.

**Comment by whichxjy** [comment]
I'm trying to compare DOM trees in this way:

``` rust
fn is_equal(lhs: &NodeRef, rhs: &NodeRef) -> bool {
    if lhs.data() != rhs.data() {
        return false;
    }

    let lhs_children = lhs.children().collect::<Vec<NodeRef>>();
    let rhs_children = rhs.children().collect::<Vec<NodeRef>>();

    if lhs_children.len() != rhs_children.len() {
        return false;
    }

    for (lc, rc) in lhs_children.iter().zip(rhs_children.iter()) {
        if !is_equal(lc, rc) {
            return false;
        }
    }

    true
}
```

But it isn't working because whitespace characters that are outside of HTML elements will be [treated as text nodes](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Whitespace) (`kuchiki` will treat newline character as `NodeRef(Text(RefCell { value: "\n" })`).

- If we want to do the DOM-based comparison, we need to handle these text nodes in DOM trees.

- If we want to do the string-based comparison, we need to ensure the generated HTML files and the expected HTML files have the same format.

**Comment by annevk** [comment]
DOM-based comparison seems a little more flexible so you can document in code what kind of whitespace or attribute differences are acceptable. If you do string-based you might run into issues such as the hashmap ordering of attributes being different.

**Comment by whichxjy** [comment]
The DOM-based comparison is working now. See #12.
